lib/twitter/twits.ex

```diff
@@ -0,0 +1,145 @@
+    from p in query, where: p.user_id == ^user_id
+  end
+
+  @doc """
+  Returns the list of posts. This fn is used for the current user.
+  ## Examples
+      iex> list_posts()
+      [%Post{}, ...]
+  """
+  @spec list_posts(User.t()) :: list(Post.t())
+  def list_posts(%Accounts.User{} = user) do
+    Post
+    |> user_posts_query(user)
+    |> Repo.all()
+  end
+
+  @doc """
+  list posts by author id
+  """
+  @spec list_posts_by_author_id(non_neg_integer() | String.t()) :: list(Post.t())
+  def list_posts_by_author_id(author_id) do
+    Repo.all(from p in Post, where: p.user_id == ^author_id)
+  end
+
+  @doc """
+  Gets a single post: takes user and the post id
+  Raises `Ecto.NoResultsError` if the Post does not exist.
+  """
+  @spec get_post!(User.t(), non_neg_integer() | String.t()) :: Post.t()
+  def get_post!(%Accounts.User{} = user, id) do
+    Post
+    |> user_posts_query(user)
+    |> Repo.get!(id)
+  end
+
+  @doc """
+  Creates a post. current_user id is registered automatically
+  ## Examples
+      iex> create_post(%{field: value})
+      {:ok, %Post{}}
+      iex> create_post(%{field: bad_value})
+      {:error, %Ecto.Changeset{}}
+  """
+  @spec create_post(User.t(), map()) :: {:ok, Post.t()} | {:error, Ecto.Changeset.t(Post.t())}
+  def create_post(%Accounts.User{} = user, attrs \\ %{}) do
+    %Post{}
+    |> Post.changeset(attrs)
+    |> Ecto.Changeset.put_assoc(:user, user)
+    |> Repo.insert()
+  end
+
+  @doc """
+  Updates a post.
+  ## Examples
+      iex> update_post(post, %{field: new_value})
+      {:ok, %Post{}}
+      iex> update_post(post, %{field: bad_value})
+      {:error, %Ecto.Changeset{}}
+  """
+  @spec update_post(User.t(), non_neg_integer() | String.t(), map()) ::
+          {:ok, Post.t()} | {:error, Ecto.Changeset.t(Post.t())}
+  def update_post(current_user, id, attrs) do
+    post = get_post!(current_user, id)
+
+    post
+    |> Post.changeset(attrs)
+    |> Repo.update()
+  end
+
+  @doc """
+  Deletes a post.
+  ## Examples
+      iex> delete_post(post)
+      {:ok, %Post{}}
+      iex> delete_post(post)
+      {:error, %Ecto.Changeset{}}
+  """
+  @spec delete_post(User.t(), non_neg_integer() | String.t()) ::
+          {:ok, Post.t()} | {:error, Ecto.Changeset.t(Post.t())}
+  def delete_post(current_user, id) do
+    post = get_post!(current_user, id)
+    Repo.delete(post)
+  end
+
+  ############# LIKES ###########################
+
+  defp likes_of_post_query(query, post_id) do
+    from l in query, where: l.post_id == ^post_id
+  end
+
+  defp get_like_query(query, post_id, %Accounts.User{id: user_id}) do
+    from l in query, where: l.post_id == ^post_id and l.user_id == ^user_id
+  end
+
+  @doc """
+  Returns the likes of a post.
+  """
+  @spec list_likes_by_post_id(non_neg_integer() | String.t()) :: list(Like.t())
+  def list_likes_by_post_id(post_id) do
+    Like
+    |> likes_of_post_query(post_id)
+    |> Repo.all()
+  end
```

Comment:
I think the design here is a bit unusual. Instead of having a separate function to get likes and a separate endpoint, I think the details of a post could have this information , both in the context (using a preload) and in the controller.
There is no hard-rule about how to manage relations, and there are some cases where you don't want to do this too (e.g. preload everything associated to a user is a trap, since everything belongs to a user). But in this case, it feels like a more natural design.

Reply:
ok!!